rfcs/0046-versioning-changes-for-hard-fork.md

```diff
@@ -0,0 +1,152 @@
+change for a given versioned type.
+
+For backwards-compatibility, we can add the annotation
+```
+ [@@@with_all_version_tags]
+```
+to a `Stable` module. For each versioned type module `Vn`, generate a
+module:
+```ocaml
+ module With_all_version_tags = struct
+   type t = ...
+   
+   let bin_read_t = ... bin_read_t ...
+   let bin_write_t = ... bin_write_t ...
+   ...
+ end
+```
+where the shadowing functions are generated as they are today. For
+this case, the constituent types must also have the same annotation,
+which is enforced by the construction of the type `t`: the constituent
+types of `t` are of the form `M.Stable.Vn.With_all_version_tags.t`.
+
+Another available annotation can be
+```
+ [@@@with_top_version_tag]
+```
+to generate:
+```ocaml
+ module With_top_version_tag = struct
+   type t = ...
+   
+   let bin_read_t = ... bin_read_t ...
+   let bin_write_t = ... bin_write_t ...
+   ...
+ end
+```
+where the shadowing functions handle a single version tag to the
+serializations of outermost type instances. The constituent types need
+not have the annotation: the type `t` will be identical to the type
+`t` in the containing `Vn` module.
+
+The existing versioning system generates a function in `Stable` modules:
+```ocaml
+ val bin_read_to_latest_opt : Bin_prot.Common.buf -> pos_ref:(int ref) -> Stable.Latest.t option
+```
+That function deserializes data by dispatching on version tags.
+Because the default serialization does not contain version tags, it no
+longer makes sense to generate such a function in `Stable`. Instead,
+generate it inside `With_all_version_tags` and `With_top_version_tag`,
+if they're created.
+
+### Version linting changes
+
+The CI version linter looks for changes to versioned types in a PR
+branch by comparing those types with counterparts in the base
+branch. The linter will complain if a new version is added, and later
+modified, even if there are no deployed nodes using the new version.
+
+Instead, the linter should use a release branch for comparison. It
+should be enough to change the script `ci_diff_types.sh` to use a
+release branch instead of `$BASE_BRANCH_NAME`. There may need to be a
+mechanism to automate that change when releases become available,
+which is beyond the scope of this RFC. The same change should be made
+to `compare_ci_diff_binables.sh`, which looks for changes to `Binable`
+functors applied in version-type modules.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The existing versioning mechanism is working acceptably well. The only
+drawback of making changes to the version-tagging is the effort to
+update `ppx_version`. That should not be a major effort, since it's
+mostly reorganizing existing code.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+If these changes are not made, network performance will be less than it
+could be.
+
+We could omit the top-tagging mechanism, since there isn't a specific
```

Comment:
The top-tagging mechanism does seem relevant for the persistent frontier information, including the persisted root.